content/automate/ManageIQ/Infrastructure/VM/Provisioning/Placement.class/__methods__/microsoft_best_fit_least_utilized.rb

```diff
@@ -1,45 +1,78 @@
 #
+module ManageIQ
+  module Automate
+    module Infrastructure
+      module VM
+        module Provisioning
+          module Placement
+            class MicrosoftBestFitLeastUtilized
+              def initialize(handle = $evm)
+                @handle = handle
+              end
 
-def clear_host(prov)
-  prov.set_option(:placement_host_name, [nil, nil])
-end
+              def main
+                @handle.log("info", "vm=[#{vm.name}], space required=[#{vm.provisioned_storage}]")
+                best_fit_least_utilized
+              end
+
+              private
+
+              def request
+                @request ||= @handle.root["miq_provision"].tap do |req|
+                  raise 'miq_provision not specified' if req.nil?
+                end
+              end
+
+              def vm
+                @vm ||= request.vm_template.tap do |vm|
+                  raise 'VM Object not specified' if vm.nil?
+                end
+              end
+
+              def storage_profile_id
+                @storage_profile_id ||= request.get_option(:placement_storage_profile).tap do |sp|
+                  @handle.log("info", "Selected storage_profile_id: #{sp}") if sp
+                end
+              end
+
+              def clear_host
+                request.set_option(:placement_host_name, [nil, nil])
+              end
+
+              def best_fit_least_utilized
+                host = storage = min_registered_vms = nil
+                request.eligible_hosts.select { |h| !h.maintenance && h.power_state == "on" }.each do |h|
+                  next if min_registered_vms && h.vms.size >= min_registered_vms
 
-prov = $evm.root["miq_provision"]
-vm   = prov.vm_template
-raise "VM not specified" if vm.nil?
-
-ems = vm.ext_management_system
-raise "EMS not found for VM [#{vm.name}]" if ems.nil?
-
-# Log space required
-$evm.log("info", "vm=[#{vm.name}], space required=[#{vm.provisioned_storage}]")
-
-host = storage = nil
-min_registered_vms = nil
-prov.eligible_hosts.each do |h|
-  next if h.maintenance
-  next unless h.power_state == "on"
-  nvms = h.vms.length
-
-  # Setting the host to filter eligible storages
-  prov.set_host(h)
-
-  if min_registered_vms.nil? || nvms < min_registered_vms
-    # Filter out storages that do not have enough free space for the Vm
-    storages = prov.eligible_storages.find_all { |s| s.free_space > vm.provisioned_storage }
-    s = storages.max_by(&:free_space)
-    unless s.nil?
-      host    = h
-      storage = s
-      min_registered_vms = nvms
+                  # Setting the host to filter eligible storages
+                  request.set_host(h)
+
+                  # Filter out storages that do not have enough free space for the Vm
+                  storages = request.eligible_storages.find_all { |s| s.free_space > vm.provisioned_storage }
+                  storages.select! { |s| s.storage_profiles.pluck(:id).include?(storage_profile_id) } if storage_profile_id
```

Comment:
We are adding the checking of storage profile.

Reply:
I removed this, thanks

Reply:
We can add it in a separate PR if it is required to check against storage profile as this PR is meant for refactoring.

Reply:
ok, thanks